cpp/src/quantiles/tdigest/tdigest_aggregation.cu

```diff
@@ -1021,6 +1021,90 @@ struct group_key_func {
+  cols.push_back(std::make_unique<cudf::column>(
+    data_type{type_id::FLOAT64},
+    total_merged_centroids,
+    rmm::device_buffer{sizeof(double) * total_merged_centroids, stream, temp_mr},
+    rmm::device_buffer{0, stream, temp_mr},
+    0));
+  cols.push_back(std::make_unique<cudf::column>(
+    data_type{type_id::FLOAT64},
+    total_merged_centroids,
+    rmm::device_buffer{sizeof(double) * total_merged_centroids, stream, temp_mr},
+    rmm::device_buffer{0, stream, temp_mr},
+    0));
+  auto result = std::make_unique<cudf::table>(std::move(cols));
+
+  // each group represents a collection of tdigest columns. each row is 1 tdigest.
+  // within each group, we want to sort all the centroids within all the tdigests
+  // in that group, using the means as the key. the "outer offsets" represent the indices of the
+  // tdigests, and the "inner offsets" represents the list of centroids for a particular tdigest.
+  //
+  //  rows
+  //  ----        centroid 0 ---------
+  //  tdigest 0   centroid 1
+  //  ----        centroid 2  group 0
+  //  tdigest 1   centroid 3
+  //  ----        centroid 4 ---------
+  //  tdigest 2   centroid 5
+  //  ----        centroid 6  group 1
+  //  tdigest 3   centroid 7
+  //              centroid 8
+  //  ----        centroid 9 --------
+  auto inner_offsets    = tdv.centroids().offsets();
+  auto centroid_offsets = cudf::detail::make_counting_transform_iterator(
+    0,
+    cuda::proclaim_return_type<size_type>(
+      [group_offsets, inner_offsets = tdv.centroids().offsets().begin<size_type>()] __device__(
+        size_type i) { return inner_offsets[group_offsets[i]]; }));
+
+  // perform the sort using the means as the key
+  size_t temp_size;
+  CUDF_CUDA_TRY(
+    cub::DeviceSegmentedSort::SortPairs(nullptr,
+                                        temp_size,
+                                        tdv.means().begin<double>(),
+                                        result->get_column(0).mutable_view().begin<double>(),
+                                        tdv.weights().begin<double>(),
+                                        result->get_column(1).mutable_view().begin<double>(),
+                                        result->num_rows(),
+                                        num_groups,
+                                        centroid_offsets,
+                                        centroid_offsets + 1,
+                                        stream.value()));
+
+  rmm::device_buffer temp_mem(temp_size, stream, temp_mr);
+  CUDF_CUDA_TRY(
+    cub::DeviceSegmentedSort::SortPairs(temp_mem.data(),
+                                        temp_size,
+                                        tdv.means().begin<double>(),
+                                        result->get_column(0).mutable_view().begin<double>(),
+                                        tdv.weights().begin<double>(),
+                                        result->get_column(1).mutable_view().begin<double>(),
```

Comment:
Calling `mutable_view` multiple times is inefficient. As mentioned above, using `device_uvector` will simplify thing a lot.

Reply:
Agreed, we can directly operate on `device_uvectors` and populate them into the `result` table at the end.